Rework onAppointmentFormOpening (#1034)
DevExpress/devextreme-documentation: api-reference/10 UI Widgets/dxScheduler/1 Configuration/onAppointmentFormOpening.md

Model data. Available only if you use Knockout.

##### field(e.popup): dxPopup
The instance of the popup that contains the form.

---
The appointment details form contains the [Form](/concepts/05%20Widgets/Form/00%20Overview.md '/Documentation/Guide/Widgets/Form/Overview/') widget whose instance is passed to this function in the **form** field. Use the [Form API](/Documentation/ApiReference/UI_Widgets/dxForm/) to customize the appointment details form. 

The form items are organized into two groups:

<table class="dx-table">
    <tr>
        <th>Group name</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>mainGroup</td>
        <td>Contains form fields that define appointment main parameters (subject, start and end dates, etc.).</td>
    </tr>
    <tr>
        <td>recurrenceGroup</td>
        <td>Contains form fields that define appointment recurrence parameters.</td>
    </tr> 
</table>

You can add a custom item to any group or create an ungrouped item and display it under the groups, as shown in the following image:

![DevExtreme Scheduler onAppointmentFormOpening](/images/UiWidgets/Scheduler_onAppointmentFormOpening.png)

> The code below adds a new form item (`phone`) to the `mainGroup` and creates an ungrouped item (`location`). Note that the array of [form items](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxForm/Configuration/#items) should be checked to ensure that it does not already contain an item with the same data field. 

Reviewer: ```suggestion
The code below adds a new form item (`phone`) to the `mainGroup` and creates an ungrouped item (`location`). Note that the array of [form items](/Documentation/ApiReference/UI_Widgets/dxForm/Configuration/#items) should be checked to ensure that it does not already contain an item with the same data field. 
```

Model data. Available only if you use Knockout.

##### field(e.popup): dxPopup
The instance of the popup that contains the form.

---

> The appointment details form is displayed inside a popup. These elements are implemented by the [Form](/Documentation/Guide/Widgets/Form/Overview/) and [Popup](/Documentation/Guide/Widgets/Popup/Overview/) widgets. Their instances are passed to **onAppointmentFormOpening** in the **form** and **popup** fields. Use the [Form API](/Documentation/ApiReference/UI_Widgets/dxForm/Configuration/) and [Popup API](/Documentation/ApiReference/UI_Widgets/dxPopup/Configuration/) to customize the corresponding widget.

Reviewer 2: ```suggestion
The appointment details form is displayed inside a popup. These elements are implemented by the [Form](/Documentation/Guide/Widgets/Form/Overview/) and [Popup](/Documentation/Guide/Widgets/Popup/Overview/) widgets. Their instances are passed to the **onAppointmentFormOpening** function in the **form** and **popup** fields. Use the [Form API](/Documentation/ApiReference/UI_Widgets/dxForm/Configuration/) and [Popup API](/Documentation/ApiReference/UI_Widgets/dxPopup/Configuration/) to customize the corresponding widget.
```

> The form items are organized into two groups:

Reviewer 2: ```suggestion
Form items are organized into two groups:
```

<table class="dx-table">
    <tr>
        <th>Group name</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>mainGroup</td>
        <td>Contains form fields that define main appointment parameters (subject, start and end dates, etc.).</td>
    </tr>
    <tr>
        <td>recurrenceGroup</td>
        <td>Contains form fields that define appointment recurrence parameters.</td>
    </tr> 
</table>

> You can add a custom item to any group or create an ungrouped item and display it under the groups, as shown in the following image:

Reviewer 2: What do you mean by display an ungrouped item under groups? ("groups" sounds like this ungrouped item is displayed under several different groups.) Please clarify.

Author: This ungrouped item is displayed under two groups.
![Scheduler_onAppointmentFormOpening](https://user-images.githubusercontent.com/66619750/88156748-85884780-cc12-11ea-92ae-ad1008c4f4ef.png)

Reviewer 2: Thank you for the clarification.

Model data. Available only if you use Knockout.

##### field(e.popup): dxPopup
The instance of the popup that contains the form.

---
The appointment details form contains the [Form](/concepts/05%20Widgets/Form/00%20Overview.md '/Documentation/Guide/Widgets/Form/Overview/') widget whose instance is passed to this function in the **form** field. Use the [Form API](/Documentation/ApiReference/UI_Widgets/dxForm/) to customize the appointment details form. 

The form items are organized into two groups:

<table class="dx-table">
    <tr>
        <th>Group name</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>mainGroup</td>

>         <td>Contains form fields that define appointment's main parameters (subject, start and end dates, etc.).</td>

Reviewer: ```suggestion
        <td>Contains form fields that define main appointment parameters (subject, start and end dates, etc.).</td>
```

    </tr>
    <tr>
        <td>recurrenceGroup</td>

>         <td>Contains form fields that define appointment's recurrence parameters.</td>

Reviewer: ```suggestion
        <td>Contains form fields that define appointment recurrence parameters.</td>
```

![DevExtreme Scheduler onAppointmentFormOpening](/images/UiWidgets/Scheduler_onAppointmentFormOpening.png)

The code below adds a new form item (`author`) to the `mainGroup` and creates an ungrouped item (`checkList`). The `mainGroup` consists of two columns. To make a  custom item span them both, set its [colSpan](/Documentation/ApiReference/UI_Widgets/dxForm/Item_Types/SimpleItem/#colSpan) to 2. Apply the same setting to an ungrouped item if it should span the `mainGroupd` and `recurrenceGroup`. Note that the array of [form items](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxForm/Configuration/#items) should be checked to ensure that it does not already contain an item with the same data field. 

---
##### jQuery

    <!-- tab: index.js -->
    $(function() {
        $("#schedulerContainer").dxScheduler({
            dataSource: [{
                text: "Start the project",
                startDate: new Date(2020, 4, 24, 9, 10),
                endDate: new Date(2020, 4, 24, 11, 20),
            }],
            currentDate: new Date(2020, 4, 24),

            onAppointmentFormOpening: function(e) {

>                 let form = e.form;

Reviewer: ```suggestion
                const form = e.form;
```

        });
    });

##### Angular

    <!-- tab: app.component.html -->
    <dx-scheduler
        [dataSource]="schedulerData"
        [currentDate]="currentDate"
        (onAppointmentFormOpening)="onAppointmentFormOpening($event)">
    </dx-scheduler>

    <!-- tab: app.component.ts -->
    import { DxSchedulerModule } from "devextreme-angular";
    // ...
    export class AppComponent {
        schedulerData = [{
            text: "Start the project",
            startDate: new Date(2020, 4, 24, 9, 10),
            endDate: new Date(2020, 4, 24, 11, 20),
        }];
        currentDate = new Date(2020, 4, 24);
        
        onAppointmentFormOpening(e) {
            const form = e.form;
            let mainGroupItems = form.itemOption('mainGroup').items; 

>             if (!mainGroupItems.find(function(i) { return i.dataField === "author" })) {

Reviewer: ```suggestion
            if (!mainGroupItems.find(function(i) { return i.dataField === "phone" })) {
```

                mainGroupItems.push({
                    colSpan: 2, 
                    label: { text: "Phone Number" },
                    editorType: "dxTextBox",
                    dataField: "phone"
                });
                form.itemOption('mainGroup', 'items', mainGroupItems);
            }
    
            let formItems = form.option("items"); 

>             if (!formItems.find(function(i) { return i.dataField === "checkList" })) {

Reviewer: ```suggestion
            if (!formItems.find(function(i) { return i.dataField === "location" })) {
```

        // ...
    })

##### Vue

    <!-- tab: App.vue -->
    <template>
        <DxScheduler

>             :dataSource="schedulerData"

Reviewer: ```suggestion
            :data-source="schedulerData"
```

>             :currentDate="currentDate"

Reviewer: ```suggestion
            :current-date="currentDate"
```

>             :on-appointment-form-opening="onAppointmentFormOpening">

Reviewer: ```suggestion
            @appointment-form-opening="onAppointmentFormOpening"
```

>         </DxScheduler>

Reviewer: ```suggestion
        />
```

    </template>

    <script>
    import 'devextreme/dist/css/dx.common.css';
    import 'devextreme/dist/css/dx.light.css';

    import DxScheduler from 'devextreme-vue/scheduler';

    export default {
        components: {

>             DxScheduler,

Reviewer: ```suggestion
            DxScheduler
```

        },
        data() {
            return {
                schedulerData: [{
                    text: "Start the project",
                    startDate: new Date(2020, 4, 24, 9, 10),
                    endDate: new Date(2020, 4, 24, 11, 20),
                }],
                currentDate: new Date(2020, 4, 24)
            }
        },
        methods: {
            onAppointmentFormOpening(e) {
                const form = e.form;
                let mainGroupItems = form.itemOption('mainGroup').items; 

>                 if (!mainGroupItems.find(function(i) { return i.dataField === "author" })) {

Reviewer: ```suggestion
                if (!mainGroupItems.find(function(i) { return i.dataField === "phone" })) {
```

                    mainGroupItems.push({
                        colSpan: 2, 
                        label: { text: "Phone Number" },
                        editorType: "dxTextBox",
                        dataField: "phone"
                    });
                    form.itemOption('mainGroup', 'items', mainGroupItems);
                }
        
                let formItems = form.option("items"); 

>                 if (!formItems.find(function(i) { return i.dataField === "checkList" })) {

Reviewer: ```suggestion
                if (!formItems.find(function(i) { return i.dataField === "location" })) {
```

                    formItems.push({
                        colSpan: 2,
                        label: { text: "Location" },
                        editorType: "dxTextBox",
                        dataField: "location"
                    });
                    form.option("items", formItems);
                }
            }
        }
    }
    </script>

##### React

    <!-- tab: App.js -->
    import React from 'react';

    import 'devextreme/dist/css/dx.common.css';
    import 'devextreme/dist/css/dx.light.css';

    import Scheduler from 'devextreme-react/scheduler';

    class App extends React.Component {
        currentDate = new Date(2020, 4, 24);
        
        onAppointmentFormOpening(e) {
            const form = e.form;
            let mainGroupItems = form.itemOption('mainGroup').items;

>             if (!mainGroupItems.find(function(i) { return i.dataField === "author" })) {

Reviewer: ```suggestion
            if (!mainGroupItems.find(function(i) { return i.dataField === "phone" })) {
```

                mainGroupItems.push({
                    colSpan: 2, 
                    label: { text: "Phone Number" },
                    editorType: "dxTextBox",
                    dataField: "phone"
                });
                form.itemOption('mainGroup', 'items', mainGroupItems);
            }
    
            let formItems = form.option("items"); 

>             if (!formItems.find(function(i) { return i.dataField === "checkList" })) {

Reviewer: ```suggestion
            if (!formItems.find(function(i) { return i.dataField === "location" })) {
```

                formItems.push({
                    colSpan: 2,
                    label: { text: "Location" },
                    editorType: "dxTextBox",
                    dataField: "location"
                });
                form.option("items", formItems);
            }
        }

        render() {
            return (
                <Scheduler 

>                     // dataSource={this.dataSource}

Reviewer: ```suggestion
```

>                     dataSource={dataSource}

Reviewer: `dataSource` is missing

>                     currentDate={this.currentDate}

Reviewer: ```suggestion
                    defaultCurrentDate={this.currentDate}
```

            if (!mainGroupItems.find(function(i) { return i.dataField === "phone" })) {
                mainGroupItems.push({
                    colSpan: 2, 
                    label: { text: "Phone Number" },
                    editorType: "dxTextBox",
                    dataField: "phone"
                });
                form.itemOption('mainGroup', 'items', mainGroupItems);
            }
    
            let formItems = form.option("items"); 
            if (!formItems.find(function(i) { return i.dataField === "location" })) {
                formItems.push({
                    colSpan: 2,
                    label: { text: "Location" },
                    editorType: "dxTextBox",
                    dataField: "location"
                });
                form.option("items", formItems);
            }
        }

        render() {
            return (
                <Scheduler 
                    dataSource={dataSource}
                    defaultCurrentDate={this.currentDate}
                    onAppointmentFormOpening={this.onAppointmentFormOpening}
                />
            );
        }
    }
    export default App;

---

> The `mainGroup` consists of two columns. To make a custom item span them both, set its [colSpan](/Documentation/ApiReference/UI_Widgets/dxForm/Item_Types/SimpleItem/#colSpan) to 2 as done in the code above. Apply the same setting to an ungrouped item if it should span the `mainGroup` and `recurrenceGroup`. 

Reviewer 2: ```suggestion
The `mainGroup` consists of two columns. To make a custom item span both columns, set its [colSpan](/Documentation/ApiReference/UI_Widgets/dxForm/Item_Types/SimpleItem/#colSpan) to '2' - as shown in the code above. Apply the same setting to an ungrouped item if it should span the `mainGroup` and `recurrenceGroup`. 
```

Author: Applied all these changes except for '2' - left it without quotes.

Reviewer 2: Okay, thank you.
